From 09a8f8759337a226402acda9687424a59cf75d88 Mon Sep 17 00:00:00 2001
From: Nick Booker <NMBooker@gmail.com>
Date: Wed, 28 Nov 2012 19:13:16 +0000
Subject: [PATCH 1/7] Added icon that changes current path to dropped-on
 directory.

---
 gtk/gtkfilechooserdefault.c |  220 +++++++++++++++++++++++++++++++++++++++++++
 gtk/gtkfilechooserprivate.h |    7 ++
 2 files changed, 227 insertions(+)

diff --git a/gtk/gtkfilechooserdefault.c b/gtk/gtkfilechooserdefault.c
index 7559d2d..80f9da2 100644
--- a/gtk/gtkfilechooserdefault.c
+++ b/gtk/gtkfilechooserdefault.c
@@ -21,6 +21,7 @@
 
 #include "gtkfilechooserdefault.h"
 
+#include "gtkalignment.h"
 #include "gtkbindings.h"
 #include "gtkcelllayout.h"
 #include "gtkcellrendererpixbuf.h"
@@ -29,6 +30,7 @@
 #include "gtkclipboard.h"
 #include "gtkcomboboxtext.h"
 #include "gtkentry.h"
+#include "gtkeventbox.h"
 #include "gtkexpander.h"
 #include "gtkfilechooserprivate.h"
 #include "gtkfilechooserdialog.h"
@@ -70,6 +72,7 @@
 #include "gtkbox.h"
 #include "gtkorientable.h"
 #include "gtkintl.h"
+#include <gdk/gdk.h>
 
 #include <errno.h>
 #include <string.h>
@@ -78,6 +81,15 @@
 #include <sys/types.h>
 #include <locale.h>
 
+static GdkAtom XdndDirectSave;
+static GdkAtom text_plain;
+static GdkAtom xa_string;
+enum {
+  GTK_XDS_SAVED,		/**< Done the save - no problem */
+  GTK_XDS_SAVE_ERROR,		/**< Error during save - reported */
+  GTK_XDS_NO_HANDLER,		/**< Used internally (sanity checking) */
+};
+
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
@@ -497,6 +509,10 @@ _gtk_file_chooser_default_class_init (GtkFileChooserDefaultClass *class)
   GtkBindingSet *binding_set;
   int i;
 
+  XdndDirectSave = gdk_atom_intern ("XdndDirectSave0", FALSE);
+  text_plain = gdk_atom_intern ("text/plain", FALSE);
+  xa_string = gdk_atom_intern ("STRING", FALSE);
+
   gobject_class->finalize = gtk_file_chooser_default_finalize;
   gobject_class->constructor = gtk_file_chooser_default_constructor;
   gobject_class->set_property = gtk_file_chooser_default_set_property;
@@ -4568,6 +4584,207 @@ location_entry_create (GtkFileChooserDefault *impl)
   gtk_entry_set_activates_default (GTK_ENTRY (impl->location_entry), TRUE);
 }
 
+/* Longest possible XdndDirectSave0 property value */
+#define XDS_MAXURILEN 4096
+
+static guchar *
+read_xds_property (GdkDragContext *context, gboolean delete)
+{
+  GdkWindow *source_window;
+  guchar  *prop_text;
+  gint	  length;
+  gchar  *retval = NULL;
+  
+  g_return_val_if_fail (context != NULL, NULL);
+
+  source_window = gdk_drag_context_get_source_window(context);
+
+  if (gdk_property_get (source_window, XdndDirectSave, text_plain,
+		       0, XDS_MAXURILEN, delete,
+		       NULL, NULL, &length, &prop_text)
+	    && prop_text)
+  {
+    /* Terminate the string */
+    retval = g_realloc (prop_text, length + 1);
+    retval[length] = '\0';
+  }
+
+  return retval;
+}
+
+static void
+write_xds_property (GdkDragContext *context, const guchar *value)
+{
+  GdkWindow *source_window = gdk_drag_context_get_source_window(context);
+  if (value)
+    {
+      gdk_property_change (source_window, XdndDirectSave,
+			   text_plain, 8, GDK_PROP_MODE_REPLACE,
+			   value, strlen ((char *) value));
+    }
+  else
+    gdk_property_delete (source_window, XdndDirectSave);
+}
+
+#define GTK_TARGET_XDS 0x584453
+static void drag_end (GtkWidget *widget, GdkDragContext *context, GtkFileChooserDefault *impl)
+{
+  g_return_if_fail (widget != NULL);
+  //g_return_if_fail (GTK_IS_SAVEBOX (widget));
+  g_return_if_fail (context != NULL);
+  printf("IN drag_end\n");
+
+  if (GTK_FILE_CHOOSER_DEFAULT (impl)->using_xds)
+    {
+      guchar  *uri;
+      uri = read_xds_property (context, TRUE);
+
+      if (uri) {
+        const gchar  *path;
+
+        /* path = rox_path_get_local (uri); */
+
+        printf("  setting uri...\n");
+        gtk_file_chooser_set_uri(GTK_FILE_CHOOSER (impl), uri);
+        
+        /* g_signal_emit (widget, savebox_signals[SAVED_TO_URI], 0,
+                   path ? path : (const gchar *) uri);
+        */
+        g_free(uri);
+
+        /* gtk_widget_destroy (widget); */
+
+        return;
+      }
+    }
+  else
+      write_xds_property (context, NULL);
+
+  /*
+  if (GTK_FILE_CHOOSER_DEFAULT (widget)->xds_data_sent)
+  {
+    g_signal_emit (widget, savebox_signals[SAVED_TO_URI], 0, NULL);
+    gtk_widget_destroy (widget);
+  }
+  */
+}
+
+static void
+button_press_over_icon (GtkImage *icon, GdkEvent *event, GtkFileChooserDefault *impl)
+{
+  const gchar *uri = NULL;
+  const gchar *leafname;
+  GdkDragContext *context;
+  uri = gtk_entry_get_text (GTK_ENTRY (impl->location_entry));
+  if (uri && *uri) {
+    leafname = g_path_get_basename (uri);
+    // confirms that we get the leafname correctly:
+    // printf ("leafname: %s\n", leafname);
+  } else {
+    return;
+    // Not sure what to do here.  Possibly show error?
+  }
+
+  impl->using_xds = FALSE;
+  impl->xds_data_sent = FALSE;
+  context = gtk_drag_begin (GTK_WIDGET (icon), impl->save_drag_targets,
+                            GDK_ACTION_COPY, event->button.button, event);
+
+  write_xds_property (context, (guchar *) leafname);
+
+  /*
+   * gtk_drag_set_icon_pixbuf (context,
+                            gtk_image_get_pixbuf (
+                              GTK_IMAGE (impl->save_drag_icon)),
+                            event->button.x, event->button.y);
+   */
+
+}
+
+static void
+drag_data_get (GtkWidget	*widget,
+	       GdkDragContext   *context,
+	       GtkSelectionData *selection_data,
+               guint            info,
+               guint32          time,
+               GtkFileChooserDefault *impl)
+{
+  GtkFileChooserDefault *savebox;
+  guchar      to_send = 'E';
+  gchar	      *uri;
+  const gchar *pathname;
+  printf("IN drag_data_get\n");
+
+  g_return_if_fail (widget != NULL);
+  //g_return_if_fail (IS_GTK_FILE_CHOOSER (widget));
+  g_return_if_fail (context != NULL);
+  g_return_if_fail (selection_data != NULL);
+
+  //savebox = (struct _GtkFileChooserDefault *) widget;
+  savebox = impl;
+
+  /* We're only concerned with the XDS protocol. Responding to other requests
+   * (including application/octet-stream) is the job of the application.
+   */
+  if (info != GTK_TARGET_XDS)
+  {
+    /* Assume that the data will be/has been sent */
+    savebox->xds_data_sent = TRUE;
+    return;
+  }
+
+  uri = read_xds_property (context, FALSE);
+  printf("uri = %s\n", uri);
+
+  to_send = 'S';
+  g_free (uri);
+  if (to_send != 'E')
+    savebox->using_xds = TRUE;
+  gtk_selection_data_set (selection_data, xa_string, 8, &to_send, 1);
+}
+
+
+/* Sets up drag-and-drop saving */
+
+static void
+save_drag_and_drop_create (GtkFileChooserDefault *impl,
+                           GtkWidget *vbox)
+{
+  GtkWidget *alignment;
+  GtkTargetEntry targets[] = { {"XdndDirectSave0", 0, GTK_TARGET_XDS} };
+  impl->save_drag_targets = gtk_target_list_new(targets,
+                                sizeof (targets) / sizeof (*targets));
+
+  impl->save_drag_icon = NULL;
+
+  //alignment = gtk_alignment_new (0.5, 0.5, 0, 0);
+  impl->save_drag_box = gtk_event_box_new ();
+  gtk_container_set_border_width (GTK_CONTAINER (impl->save_drag_box), 4);
+  gtk_widget_add_events(impl->save_drag_box, GDK_BUTTON_PRESS_MASK);
+  g_signal_connect (impl->save_drag_box, "button-press-event",
+                    G_CALLBACK (button_press_over_icon), impl);
+  g_signal_connect (impl->save_drag_box, "drag-end",
+                    G_CALLBACK (drag_end), impl);
+  g_signal_connect (impl->save_drag_box, "drag-data-get",
+                    G_CALLBACK (drag_data_get), impl);
+
+ 
+  impl->save_drag_icon = gtk_image_new_from_stock (GTK_STOCK_SAVE,
+                                                   GTK_ICON_SIZE_DND);
+  //gtk_drag_source_set(GTK_WIDGET (impl->save_drag_icon), GDK_BUTTON1_MASK, NULL, 0, GDK_ACTION_COPY);
+  //gtk_drag_source_set_icon_stock(GTK_WIDGET (impl->save_drag_icon), "gtk-save");
+  //g_signal_connect (impl->save_drag_box, "button-press-event", G_CALLBACK (button_press_over_icon), impl);
+  gtk_container_add (GTK_CONTAINER (impl->save_drag_box), impl->save_drag_icon);
+  gtk_widget_show(GTK_WIDGET (impl->save_drag_icon));
+  gtk_widget_show(GTK_WIDGET (impl->save_drag_box));
+  //gtk_widget_show(GTK_WIDGET (alignment));
+  gtk_box_pack_start (GTK_BOX (vbox), impl->save_drag_box, FALSE, FALSE, 0);
+
+
+
+    //gtk_box_pack_start (GTK_BOX (vbox), GTK_WIDGET (impl->save_icon), FALSE, FALSE, 0);
+}
+
 /* Creates the widgets specific to Save mode */
 static void
 save_widgets_create (GtkFileChooserDefault *impl)
@@ -4582,6 +4799,8 @@ save_widgets_create (GtkFileChooserDefault *impl)
 
   vbox = gtk_box_new (GTK_ORIENTATION_VERTICAL, 12);
 
+  save_drag_and_drop_create (impl, vbox);
+
   impl->save_widgets_table = gtk_grid_new ();
   gtk_box_pack_start (GTK_BOX (vbox), impl->save_widgets_table, FALSE, FALSE, 0);
   gtk_widget_show (impl->save_widgets_table);
@@ -4604,6 +4823,7 @@ save_widgets_create (GtkFileChooserDefault *impl)
   gtk_widget_show (impl->location_entry);
   gtk_label_set_mnemonic_widget (GTK_LABEL (widget), impl->location_entry);
 
+
   /* Folder combo */
   impl->save_folder_label = gtk_label_new (NULL);
   gtk_widget_set_halign (impl->save_folder_label, GTK_ALIGN_START);
diff --git a/gtk/gtkfilechooserprivate.h b/gtk/gtkfilechooserprivate.h
index e51a041..076e26e 100644
--- a/gtk/gtkfilechooserprivate.h
+++ b/gtk/gtkfilechooserprivate.h
@@ -164,6 +164,13 @@ struct _GtkFileChooserDefault
 
   GtkWidget *save_folder_label;
 
+  /* Xdnd Direct Save */
+  GtkWidget *save_drag_icon;
+  GtkWidget *save_drag_box;
+  GtkTargetList *save_drag_targets;
+  gboolean using_xds;
+  gboolean xds_data_sent;
+
   /* The file browsing widgets */
   GtkWidget *browse_widgets_box;
   GtkWidget *browse_header_box;
-- 
1.7.9.5

